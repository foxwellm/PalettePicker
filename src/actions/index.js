export const setCurrentPalette = (palette) => ({
  type: 'SET_CURRENT_PALETTE',
  palette
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const removeProject = (projectId) => ({
  type: 'REMOVE_PROJECT',
  projectId
});

export const setPalettes = (palettes) => ({
  type: 'SET_PALETTES',
  palettes
});

export const addPalette = (palette) => ({
  type: 'ADD_PALETTE',
  palette
});

export const removePalette = (paletteId) => ({
  type: 'REMOVE_PALETTE',
  paletteId
});

export const setSearchProjects = (search) => ({
  type: 'SEARCH_PROJECTS',
  search
});

export const setSearching = (bool) => ({
  type: 'SEARCHING_PROJECTS',
  bool
});

export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});