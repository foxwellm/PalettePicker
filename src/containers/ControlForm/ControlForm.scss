.form-div {
  text-align: center;
}

.save-palette-div {
  display: block;
  margin: 20px auto;
}

.new-project-div {
  display: block;
  margin: 20px auto;
}