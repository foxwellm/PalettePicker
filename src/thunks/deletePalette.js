import { setLoading, setError, removePalette } from '../actions';
import { fetchData } from '../utils/api';

export const deletePalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/palettes/${id}`, 'DELETE');
      dispatch(removePalette(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}