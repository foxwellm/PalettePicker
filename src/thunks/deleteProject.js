import { setLoading, setError, removeProject } from '../actions';
import { fetchData } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/projects/${id}`, 'DELETE');
      dispatch(removeProject(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}