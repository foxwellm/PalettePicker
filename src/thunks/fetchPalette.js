import { setLoading, setError, setPalettes } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const palette = await fetchData(`/api/v1/palettes/${id}`, 'GET');
      dispatch(setPalettes([palette]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};