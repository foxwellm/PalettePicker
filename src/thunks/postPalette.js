import { setLoading, setError, addPalette } from '../actions';
import { fetchData } from '../utils/api';

export const postPalette = (palette) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const newPalette = await fetchData('/api/v1/palettes', 'POST', palette);
      dispatch(addPalette(newPalette[0]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false))
  }
}