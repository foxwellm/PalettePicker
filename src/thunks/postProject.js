import { setLoading, setError, addProject } from '../actions';
import { fetchData } from '../utils/api';

export const postProject = (name) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const newProjectId = await fetchData('/api/v1/projects', 'POST', { name });
      dispatch(addProject({ id: newProjectId[0], name }));
      return newProjectId[0]
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false))
  }
}