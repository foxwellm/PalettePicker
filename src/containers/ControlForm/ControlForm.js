import React, { Component } from 'react';
import { connect } from 'react-redux';
import { postProject } from '../../thunks/postProject';
import { postPalette } from '../../thunks/postPalette';
import { setError } from '../../actions';
import PropTypes from 'prop-types';

export class ControlForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedProject: '', 
      project: '',
      palette: ''
    };
  };

  saveProject = async () => {
    const { setError, postProject } = this.props;
    setError('');
    const id = await postProject(this.state.project);
    this.setState({ selectedProject: id.toString() });
    this.clearInput('project');
  }

  savePalette = async () => {
    const { setError, postPalette, currentPalette } = this.props;
    const { palette, selectedProject } = this.state;
    setError('');
    await postPalette({
      name: palette,
      color1: currentPalette[0],
      color2: currentPalette[1],
      color3: currentPalette[2],
      color4: currentPalette[3],
      color5: currentPalette[4],
      project_id: parseInt(selectedProject)
    });
    this.clearInput('palette');
  }

  clearInput = (name) => {
    !this.props.error && this.setState({ [name]: '' })
  };

  handleChange = (e) => {
    const { value, id } = e.target;
    this.setState({
      [id]: value
    });
  };

  changeSelectedProject = (e) => {
    this.setState({ selectedProject: e.target.value });
  };

  render() {
    const { error, projects } = this.props;
    const { project, palette, selectedProject } = this.state;
    return (
      <div className='form-div'>
        <div className='new-project-div'>
          <label>Create New Project:
            <input onChange={this.handleChange} value={project} id='project' placeholder='Add project name'></input>
          </label>
          <button onClick={this.saveProject} className='standard-button'>Save Project</button>
        </div>
        <div className='save-palette-div'>
          <select onChange={this.changeSelectedProject} value={selectedProject}>
            <option value=''>Choose a project</option>
            {
              projects.map(project => {
                return(<option value={project.id} key={project.id}>{project.name}</option>)
              })
            }
          </select>
          <input onChange={this.handleChange} value={palette} id='palette' placeholder='Add palette name'></input>
          <button onClick={this.savePalette} className='standard-button' disabled={selectedProject.length ? false : true}>Save Palette</button>
        </div>
        <p>{error}</p>
      </div>
    );
  };
};

export const mapStateToProps = (state) => ({
  error: state.error,
  currentPalette: state.currentPalette,
  projects: state.projects,
});

export const mapDispatchToProps = (dispatch) => ({
  postProject: (project) => dispatch(postProject(project)),
  setError: (error) => dispatch(setError(error)),
  postPalette: (palette) => dispatch(postPalette(palette)),
});

ControlForm.propTypes = {
  error: PropTypes.string,
  currentPalette: PropTypes.array,
  projects: PropTypes.array,
  postProject: PropTypes.func,
  setError: PropTypes.func,
  postPalette: PropTypes.func,
};

ControlForm.defaultProps = {
  error: '',
  currentPalette: [],
  projects: [],
};

export default connect(mapStateToProps, mapDispatchToProps)(ControlForm);