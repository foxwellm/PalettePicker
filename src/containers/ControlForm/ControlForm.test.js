import React from 'react';
import { ControlForm, mapStateToProps, mapDispatchToProps } from './ControlForm';
import { shallow } from 'enzyme';
import { postProject } from '../../thunks/postProject';
import { setError } from '../../actions';
import { postPalette } from '../../thunks/postPalette';
jest.mock('../../thunks/postProject');
jest.mock('../../actions');
jest.mock('../../thunks/postPalette');

describe('ControlForm', () => {
  let wrapper;
  let mockProjects;
  let mockCurrentPalette;
  let mockError;
  let mockPostProject = jest.fn();
  let mockPostPalette = jest.fn();
  let mockSetError = jest.fn();

  beforeEach(() => {
    mockError = '';
    mockProjects = [{ id: 1, name: 'Project 1' }, { id: 2, name: 'Project 2' }];
    mockCurrentPalette = ['#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff'];

    wrapper = shallow(
      <ControlForm
        projects={mockProjects}
        currentPalette={mockCurrentPalette}
        error={mockError}
        postProject={mockPostProject}
        postPalette={mockPostPalette}
        setError={mockSetError}
      />
    );
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });


  describe('saveProject', () => {

    it('should call setError with correct params', () => {
      wrapper.instance().saveProject();
      expect(mockSetError).toHaveBeenCalledWith('')
    });

    it('should call postProject with correct params', () => {
      const expectedProjectState = 'New Project';
      wrapper.setState({project: expectedProjectState});
      wrapper.instance().saveProject();
      expect(mockPostProject).toHaveBeenCalledWith(expectedProjectState);
    });

    it.skip('should call clearInput with correct params', () => {
      wrapper.instance().clearInput = jest.fn();
      wrapper.instance().saveProject();
      expect(wrapper.instance().clearInput).toHaveBeenCalledWith('project');
    });

  });

  describe('savePalette', () => {

    it('should call setError with correct params', () => {
      wrapper.instance().savePalette();
      expect(mockSetError).toHaveBeenCalledWith('');
    });

    it.skip('should call clearInput with correct params', () => {
      wrapper.instance().clearInput = jest.fn();
      wrapper.instance().saveProject();
      expect(wrapper.instance().clearInput).toHaveBeenCalledWith('palette')
    });

  });

  describe('clearInput', () => {

    it('should change state to current project', () => {
      const initialProjectState = 'Project 1';
      const expectedState = '';
      wrapper.setState({project: initialProjectState})
      expect(wrapper.state('project')).toEqual(initialProjectState);
      wrapper.instance().clearInput('project');
      expect(wrapper.state('project')).toEqual(expectedState);
    });

  });

  describe('handleChange', () => {

    it('should change state to current project', () => {
      const initialProjectState = '';
      const mockEvent = {
        target: {
          value: 'New Project',
          id: 'project'
        }
      };
      expect(wrapper.state('project')).toEqual(initialProjectState);
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('project')).toEqual(mockEvent.target.value);
    });

  });

  describe('changeSelectedProject', () => {

    it('should change state to current project', () => {
      const initialSelectedProjectState = '';
      const mockEvent = {
        target: {
          value: 'New Project'
        }
      };
      expect(wrapper.state('selectedProject')).toEqual(initialSelectedProjectState);
      wrapper.instance().changeSelectedProject(mockEvent);
      expect(wrapper.state('selectedProject')).toEqual(mockEvent.target.value);
    });

  });

  describe('mapStateToProps', () => {

    it('should return an object of props', () => {
      const mockState = {
        projects: mockProjects,
        currentPalette: mockCurrentPalette,
        error: mockError,
        otherState: false,
      };
      const expected = {
        projects: mockProjects,
        currentPalette: mockCurrentPalette,
        error: mockError,
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });

  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch when postProject is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = postProject();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.postProject();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setError is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setError();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setError();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when postPalette is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = postPalette();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.postPalette();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

  });

});