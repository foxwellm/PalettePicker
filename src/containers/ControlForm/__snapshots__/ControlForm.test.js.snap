// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlForm should match the correct snapshot 1`] = `
<div
  className="form-div"
>
  <div
    className="new-project-div"
  >
    <label>
      Create New Project:
      <input
        id="project"
        onChange={[Function]}
        placeholder="Add project name"
        value=""
      />
    </label>
    <button
      className="standard-button"
      onClick={[Function]}
    >
      Save Project
    </button>
  </div>
  <div
    className="save-palette-div"
  >
    <select
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Choose a project
      </option>
      <option
        key="1"
        value={1}
      >
        Project 1
      </option>
      <option
        key="2"
        value={2}
      >
        Project 2
      </option>
    </select>
    <input
      id="palette"
      onChange={[Function]}
      placeholder="Add palette name"
      value=""
    />
    <button
      className="standard-button"
      disabled={true}
      onClick={[Function]}
    >
      Save Palette
    </button>
  </div>
  <p />
</div>
`;
